### ISSUE TITLE
Rust-like features

### ISSUE THREAD

**Issue by ramonGonzEdu** [riicchhaarrd/ocean#7]
Another idea is adding rusty features like enums with data and better `switch` (maybe with alternate keyword for support).
This would replace something like `std::variant` and `std::visit` of c++.

**Comment by ramonGonzEdu** [comment]
Also, this is an idea on how to acheive perfect memory management (single-threaded)

Allow properties on a struct to be one of these three: direct value, reference, weak reference.

## What they do
- A - Direct value: Is just included directly in the struct
- B - Reference: A reference counted pointer that points to null_ptr or a valid address containing some value
- C - Weak Reference: A direct pointer that points to null_ptr, a valid address, or an invalid free'd address. You would need to null-check before using the pointer.

## How to make them guarantee memory safety:
- A structs will be marked with what types they reference (B) and what types they contain (A). Note that this would stack since in the scenario:
```cpp
struct A {
B b;
}
struct B {
C c;
}
struct C {
A* a;
}
```
A would be marked with both `B` which it directly contains, and `C` which it indirectly contains. It is not marked with itself, because C contains a weak ref/raw pointer (B).
- Using this system, a struct is not allowed to be marked with itself. 
- With the above example, C could not contain an A reference/value because that would keep alive the A -> B -> C which would cause a memory leak since they would hold strong references to each other. 
- In a Linked list example, each node could either strong reference the next node or the previous, but not both as that would keep each node referencing the other, thus causing them to stay alive.
- Note that although struct `A` holding a strong reference to itself is physically possible, holding a value of itself would cause an infinite recursion meaning that it would take either infinite ram or no ram.

## TLDR
A struct shouldn't be able to hold a reference-counted pointer or direct value of itself (including in ref-counted or value children), but can hold a weak/raw pointer of itself.

**Comment by ramonGonzEdu** [comment]
Note that the above could be enforced by some sort of attribute but would require built-in containers like hashmaps and a built-in linked list as a custom linked list would probably lose its head if you move up the list without keeping a separate reference to the head (can't be a reference in each node as that would also cause a reference loop)
